File: src/main/java/forpleuvoir/hiirosakura/client/feature/task/executor/base/JavaScriptInterface.java
package forpleuvoir.hiirosakura.client.feature.task.executor.base;


import forpleuvoir.hiirosakura.client.HiiroSakuraClient;
import forpleuvoir.hiirosakura.client.feature.input.AnalogInput;
import forpleuvoir.hiirosakura.client.feature.task.TimeTask;
import forpleuvoir.hiirosakura.client.feature.task.TimeTaskHandler;
import forpleuvoir.hiirosakura.client.mixin.MixinMinecraftClientInterface;
import forpleuvoir.hiirosakura.client.util.ServerInfoUtil;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.screen.ConnectScreen;
import net.minecraft.client.gui.screen.TitleScreen;
import net.minecraft.client.gui.screen.multiplayer.MultiplayerScreen;
import net.minecraft.client.network.ServerAddress;

import static forpleuvoir.hiirosakura.client.feature.input.AnalogInput.Key.*;

/**
 * js脚本接口实现
 *
 * @author forpleuvoir
 * <p>#project_name hiirosakura
 * <p>#package forpleuvoir.hiirosakura.client.feature.task.executor.base
 * <p>#class_name JavaScriptInterface
 * <p>#create_time 2021/7/27 23:20
 */
public class JavaScriptInterface implements IJavaScriptInterface {
    private static final MinecraftClient mc = MinecraftClient.getInstance();
    private static final HiiroSakuraClient hs = HiiroSakuraClient.getInstance();
    private static final AnalogInput input = AnalogInput.getInstance();

    @Override
    public void forward(int tick) {
        input.set(FORWARD, tick);
    }

    @Override
    public void back(int tick) {
        input.set(BACK, tick);
    }

    @Override
    public void left(int tick) {
        input.set(LEFT, tick);
    }

    @Override
    public void right(int tick) {
        input.set(RIGHT, tick);
    }

    @Override
    public void jump(int tick) {
        input.set(JUMP, tick);
    }

    @Override
    public void sneak(int tick) {
        input.set(SNEAK, tick);
    }

    @Override
    public void attack(int tick) {
        input.set(ATTACK, tick);
    }

    @Override
    public void use(int tick) {
        input.set(USE, tick);
    }

    @Override
    public void pickItem(int tick) {
        input.set(PICK_ITEM, tick);
    }

    @Override
    public void doAttack() {
        ((MixinMinecraftClientInterface) mc).callDoAttack();
    }

    @Override
    public void doItemUse() {
        ((MixinMinecraftClientInterface) mc).callDoItemUse();
    }

    @Override
    public void doItemPick() {
        ((MixinMinecraftClientInterface) mc).callDoItemPick();
    }

    @Override
    public void joinServer(String address) {
        joinServer(address, 5);
    }

    @Override
    public void joinServer(String address, int maxConnect) {
        if (mc.world == null && maxConnect > ServerInfoUtil.getDisConnectCounter()) {
            var multiplayerScreen = new MultiplayerScreen(new TitleScreen());
            mc.setScreen(multiplayerScreen);
            TimeTaskHandler.getInstance().addTask(
                    TimeTask.once(
                            task -> {
                                mc.disconnect();
                                ConnectScreen.connect(multiplayerScreen, mc, ServerAddress.parse(address), null);
                            },
                            15,
                            "#joinServer"
                    )
            );
        }
    }

    @Override
    public void sendChatMessage(String message) {
        hs.sendMessage(message);
    }
}

File: src/main/kotlin/forpleuvoir/hiirosakura/client/feature/task/executor/base/JavaScriptInterface.kt
package forpleuvoir.hiirosakura.client.feature.task.executor.base

import forpleuvoir.hiirosakura.client.HiiroSakuraClient
import forpleuvoir.hiirosakura.client.feature.input.AnalogInput
import forpleuvoir.hiirosakura.client.feature.input.AnalogInput.Key.*
import forpleuvoir.hiirosakura.client.feature.task.TimeTask.Companion.once
import forpleuvoir.hiirosakura.client.feature.task.TimeTaskHandler
import forpleuvoir.hiirosakura.client.feature.task.executor.SimpleExecutor
import forpleuvoir.hiirosakura.client.mixin.MixinMinecraftClientInterface
import forpleuvoir.hiirosakura.client.util.ServerInfoUtil
import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.screen.ConnectScreen
import net.minecraft.client.gui.screen.TitleScreen
import net.minecraft.client.gui.screen.multiplayer.MultiplayerScreen
import net.minecraft.client.network.ServerAddress
import net.minecraft.network.packet.c2s.play.UpdateSelectedSlotC2SPacket

/**
 * js脚本接口实现
 *
 * @author forpleuvoir
 *
 * #project_name hiirosakura
 *
 * #package forpleuvoir.hiirosakura.client.feature.task.executor.base
 *
 * #class_name JavaScriptInterface
 *
 * #create_time 2021/7/27 23:20
 */
class JavaScriptInterface : IJavaScriptInterface {
    override fun forward(tick: Int) {
        input[FORWARD] = tick
    }

    override fun back(tick: Int) {
        input[BACK] = tick
    }

    override fun left(tick: Int) {
        input[LEFT] = tick
    }

    override fun right(tick: Int) {
        input[RIGHT] = tick
    }

    override fun jump(tick: Int) {
        input[JUMP] = tick
    }

    override fun sneak(tick: Int) {
        input[SNEAK] = tick
    }

    override fun attack(tick: Int) {
        input[ATTACK] = tick
    }

    override fun use(tick: Int) {
        input[USE] = tick
    }

    override fun pickItem(tick: Int) {
        input[PICK_ITEM] = tick
    }

    override fun doAttack() {
        (mc as MixinMinecraftClientInterface).callDoAttack()
    }

    override fun doItemUse() {
        (mc as MixinMinecraftClientInterface).callDoItemUse()
    }

    override fun doItemPick() {
        (mc as MixinMinecraftClientInterface).callDoItemPick()
    }

    override fun joinServer(address: String) {
        joinServer(address, 5)
    }

    override fun joinServer(address: String, maxConnect: Int) {
        if (mc.world == null && maxConnect > ServerInfoUtil.disConnectCounter) {
            val multiplayerScreen = MultiplayerScreen(TitleScreen())
            mc.setScreen(multiplayerScreen)
            TimeTaskHandler.INSTANCE!!.addTask(
                once(
                    SimpleExecutor {
                        mc.disconnect()
                        ConnectScreen.connect(multiplayerScreen, mc, ServerAddress.parse(address), null)
                    },
                    15,
                    "#joinServer"
                )
            )
        }
    }

    override fun sendChatMessage(message: String) {
        hs.sendMessage(message)
    }

    override fun dropItem(slot: Int, all: Boolean) {
        val player = mc.player
        val inventory = player?.inventory
        player?.dropSelectedItem(all)
        inventory?.updateItems()
        mc.networkHandler?.sendPacket(UpdateSelectedSlotC2SPacket(inventory!!.selectedSlot))
    }

    override fun dropItem(name: String, all: Boolean) {
        TODO("Not yet implemented")
    }

    companion object {
        private val mc = MinecraftClient.getInstance()
        private val hs = HiiroSakuraClient
        private val input = AnalogInput
    }
}